scipy/stats/_stats_py.py

```diff
@@ -3959,52 +3959,44 @@ def f_oneway(*samples, axis=0, equal_var=True):
     # the related issue: https://github.com/scipy/scipy/issues/11122
     else:
         # calculate basic statistics for each sample
-        samples_arrs = np.array(samples)
+        means = np.asarray([np.mean(sample, axis=axis) for sample in samples])
+        ns = np.asarray(
+            [np.apply_along_axis(len, axis=axis, arr=sample) for sample in samples])
+        k = len(samples)
 
-        # adjust the shape of the data based on the arrary dimension for the calculation
-        # along with input axis
-        if samples_arrs.ndim <= 2:
-            samples_arrs = samples_arrs.T
-        else:
-            samples_arrs = samples_arrs.transpose(
-                *range(1, samples_arrs.ndim - 1), 0, -1
-            )
+        vars_ = np.asarray([np.var(sample, axis=axis, ddof=1) for sample in samples])
 
-        means = samples_arrs.mean(axis=axis)
-        ns = np.apply_along_axis(np.size, axis=axis, arr=samples_arrs)
-        k = np.apply_along_axis(np.shape, axis=axis, arr=ns)
-
-        vars_ = np.var(samples_arrs, axis=axis, ddof=1)
         # calculate weight by number of data and varianc
+        # ref.[4] p.330 ($$w = 1/{\lambda s^2}$$ where $$\lambda=1/n$$ in ref)
         ws = ns / vars_
 
         # calculate adjusted grand mean
-        mean_prime = (np.sum(ws * means, axis=axis, keepdims=True) /
-                      np.sum(ws, axis=axis, keepdims=True))
+        # ref.[4] p.330 (notation is $$\hat y$$ in ref)
+        mean_prime = np.sum(ws * means, axis=0) / np.sum(ws, axis=0)
```

Comment:
```suggestion
        mean_prime = np.vecdot(ws, means, axis=0) / np.sum(ws, axis=0)
```
Please test this locally before committing. 
This probably doesn't make much difference for only a few samples, but `vecdot(x, y)` can be much, much faster than `sum(x * y)` when there are many elements.